File: algorithms/2_medium/F/01642_furthest_building_you_can_reach/heap.cpp
class Solution {
public:
    static int furthestBuilding(const vector<int>& heights, int bricks, int ladders) {
        if (heights.empty()) { return -1; }

        const size_t len = heights.size();
        auto heap = priority_queue<int>();

        for (size_t idx = 0; idx + 1 < len; idx += 1) {
            if (heights[idx] < heights[idx + 1]) {
                const int need = heights[idx + 1] - heights[idx];
                bricks -= need;
                heap.push(need);
                if (bricks < 0) {
                    bricks += heap.top();
                    heap.pop();
                    ladders -= 1;
                    if (ladders < 0) {
                        return static_cast<int>(idx);
                    }
                }
            }
        }

        return static_cast<int>(len - 1);
    }
};

File: algorithms/2_medium/F/01642_furthest_building_you_can_reach/heap.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn furthest_building(heights: Vec<i32>, mut bricks: i32, mut ladders: i32) -> i32 {
        if heights.is_empty() { return -1; }

        let len = heights.len();
        let mut heap: BinaryHeap<i32> = BinaryHeap::new();

        for idx in 0..(len - 1) {
            if heights[idx] < heights[idx + 1] {
                let need = heights[idx + 1] - heights[idx];
                bricks -= need;
                heap.push(need);
                if bricks < 0 {
                    bricks += heap.pop().unwrap();
                    ladders -= 1;
                    if ladders < 0 {
                        return idx.try_into().unwrap();
                    }
                }
            }
        }

        (len - 1).try_into().unwrap()
    }
}
